### DISCUSSION TITLE
About keep-alive mechanism

### DISCUSSION THREAD

**Discussion by wjswjq** [eclipse-iceoryx/iceoryx#2000] [General]
Roudi use keepalive message to monitor the process status. Why not just monitor the running status of each process from system since it has all pids of them. Is there any consideration?

**Comment by elfenpiff** [comment]
@wjswjq This does not work so well. What if a process crashes and a new one comes up with the same pid? Then you may end up monitoring the wrong process. Keepalive messages are a good middle ground for the open-source version of iceoryx, in the extended closed-source version we use another mechanism that does not come with the jittering issue of unix domain sockets and has the downside of pid watching.

**Reply by wjswjq** [reply]
So maybe roudi shall check both pid and process name?

**Reply by elfenpiff** [reply]
This would minimize the problem but still, the process could crash and be restarted by some launcher with the same process id. Unlikely but possible. Also, what happens when the process has not crashed but is in some kind of deadlock or is a zombie process, dead but still listed in the process table - see: https://en.wikipedia.org/wiki/Zombie_process?

**Reply by wjswjq** [reply]
Thanks for you reply.
1. If process with the same pid is restarted, it will register to roudi. so this might not be a problem. 
2. Keepalive heartbeat does not work for process in deadlock. An extra monitor is required to stop this process.
3. Zombie process is a problem.

**Reply by wjswjq** [reply]
@elfenpiff  Furthermore, now each iceoryx process use create a file lock for IpcInterface. Can roudi just check the filelock status?

**Reply by elfenpiff** [reply]
@wjswjq This would be possible but it would be a heavyweight file operation where roudi has to list all lock files in a specific directory and needs to try to lock them. This would require some refactoring and could burn your cpu if you have several hundred processes.
